OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnPowerState.cs

```diff
@@ -0,0 +1,84 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Needs power to operate.")]
+	public class GrantConditionOnPowerStateInfo : ConditionalTraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("Condition to grant.")]
+		public readonly string Condition = null;
+
+		public readonly PowerState ValidPowerStates = PowerState.Low | PowerState.Critical;
+
+		public override object Create(ActorInitializer init) { return new GrantConditionOnPowerState(init.Self, this); }
+	}
+
+	public class GrantConditionOnPowerState : ConditionalTrait<GrantConditionOnPowerStateInfo>, INotifyOwnerChanged, INotifyPowerLevelChanged
+	{
+		PowerManager playerPower;
+		ConditionManager conditionManager;
+		int conditionToken = ConditionManager.InvalidConditionToken;
+
+		bool validPowerState;
+
+		public GrantConditionOnPowerState(Actor self, GrantConditionOnPowerStateInfo info)
+			: base(info)
+		{
+			playerPower = self.Owner.PlayerActor.Trait<PowerManager>();
+		}
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			Update(self);
+		}
+
+		protected override void TraitDisabled(Actor self)
+		{
+			Update(self);
+		}
+
+		void Update(Actor self)
+		{
+			if (conditionManager == null || IsTraitDisabled)
+				return;
+
+			validPowerState = Info.ValidPowerStates.HasFlag(playerPower.PowerState);
+
+			if (validPowerState && conditionToken == ConditionManager.InvalidConditionToken)
+				conditionToken = conditionManager.GrantCondition(self, Info.Condition);
```

Comment:
If `Update` is called and `validPowerState` has the same value that it did previously, then you will double grant or double revoke conditions.  This may be why buildings aren't returning from low power.

You should cache the last `validPowerState` and then exit early if it has not changed.